snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+        # pausing them here would be to avoid buffering messages that will
+        # likely be discarded. On the other hand, by the time we get to consume
+        # these messages, the offsets may have been synchronized at that point.
+        # It's hard to predict which scenario is more likely -- in the interest
+        # of buffering and low latency, it probably makes sense to avoid
+        # pausing here?)
+        with self.__lock:
+            # TODO: This should be limited to only streams that are currently
+            # paused due to the remote offsets -- we should even be able to get
+            # those streams without acquiring the lock here.
+            for stream, offset in self.__backend.tell().items():
+                effective_remote_offset = self.__get_effective_remote_offset(stream)
+                if (
+                    effective_remote_offset is not None
+                    and effective_remote_offset > offset
+                ):
+                    # TODO: This should be more intelligent (see comment in the
+                    # assignment callback in ``subscribe`` for more details.)
+                    self.__backend.resume([stream])
+
+        message = self.__backend.poll(timeout)
+        if message is None:
+            return None
+
+        # Check to ensure that consuming this message will not cause the local
+        # offset to exeed the remote offset(s).
+        with self.__lock:
+            effective_remote_offset = self.__get_effective_remote_offset(message.stream)
+            if (
+                effective_remote_offset is None
+                or message.offset >= effective_remote_offset
+            ):
+                # If this message does exceed the remote offset(s), it should be
+                # dropped, the stream should be paused, and the offset should be
+                # reset so that this message is consumed again when the stream is
+                # resumed.
+                self.__backend.pause([message.stream])
+                self.__backend.seek({message.stream: message.offset})
+
+        return message
+
+    def tell(self) -> Mapping[TStream, TOffset]:
+        return self.__backend.tell()
+
+    def seek(self, offsets: Mapping[TStream, TOffset]) -> None:
+        # TODO: Streams should be paused or resumed on seek depending on their
+        # relationship to the remote offset(s).
+        raise NotImplementedError
+
+    def pause(self, streams: Sequence[TStream]) -> None:
+        raise NotImplementedError
+
+    def resume(self, streams: Sequence[TStream]) -> None:
+        raise NotImplementedError
+
+    def commit(self) -> Mapping[TStream, TOffset]:
+        return self.__backend.commit()
+
+    def close(self, timeout: Optional[float] = None) -> None:
+        self.__shutdown_requested.set()
+        try:
+            return self.__backend.close(timeout)
+        finally:
+            self.__commit_log_consumer_future.result(timeout=0)
```

Comment:
I think commit-on-close is going to be moved to the higher level consumer (and out of the backend) when I get a chance, which will remove the necessity for the timeout here and generally simplify this, since it'll be a synchronous/blocking option instead of this kind of weird pairing of synchronous and asynchronous stuff going on.

Reply:
Turns out, this isn't so much related to commit-on-close as it is related to `commit` being a synchronous call that triggers an asynchronous publish when using the `KafkaConsumerBackendWithCommitLog` backend: https://github.com/getsentry/snuba/blob/7b88bed627db05b5d653c2f848461f2dd0ec4441/snuba/utils/streams/consumers/backends/kafka.py#L404-L408

`close` is conceptually treated as `join` on all of the outstanding produce calls that were initiated here: https://github.com/getsentry/snuba/blob/7b88bed627db05b5d653c2f848461f2dd0ec4441/snuba/utils/streams/consumers/backends/kafka.py#L392-L400

The better approach here would be to move the commit log publishing to the higher level consumer, instead of coupling it to the backend. That'll take work beyond this change, though, so marking this thread as resolved for now.